### PR TITLE
ci: add smoke tests integration

### PR DESCRIPTION
**Pull request by kruskall**
Initial work to integrate smoke testing scripts into CI. 

Mostly inspired by the work in the mutating webhook and APM server, add make tasks to make testing easier and provide a target for CI to discover, run and cleanup the environment.

The tasks are flexible and support muiltiple smoke tests (currently there's only one).

Related to https://github.com/elastic/apm-aws-lambda/issues/217

### CODE REVIEW

**Comment by simitt** [comment]
@kruskall could you provide a sentence or two to the description. 

@v1v could you (or someone from the team) review the ci automation part here please (does not need to be merged before 8.5 FF).

File: .ci/jobs/apm-aws-lambda-smoketests-mbp.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+          disable: false
+          recursive: true
+          parent-credentials: true
+          timeout: 100
+        timeout: '15'
+        use-author: true
+        wipe-workspace: true
+        property-strategies:
+          all-branches:
+          - suppress-scm-triggering: true
```
Comment (v1v) [inline .ci/jobs/apm-aws-lambda-smoketests-mbp.yml:44]: This section will avoid listening for SCM events, such as git merge commits or github commands, is that the expected behaviour?

File: .ci/smoke-tests.groovy [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        // cloud tags
+        TF_VAR_BUILD_ID = "${env.BUILD_ID}"
+        TF_VAR_ENVIRONMENT= 'ci'
+        TF_VAR_BRANCH = "${env.BRANCH_NAME.toLowerCase().replaceAll('[^a-z0-9-]', '-')}"
+        TF_VAR_REPO = "${REPO}"
+      }
+      steps {
+        dir ("${BASE_DIR}") {
+          withTestClusterEnv {
+            withGoEnv(version: readFile(file: ".go-version").trim()) {
```
Comment (v1v) [inline .ci/smoke-tests.groovy:52]:
```suggestion
            withGoEnv() {
```

File: .ci/smoke-tests.groovy [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+              }
+            }
+          }
+        }
+      }
+      post {
+        always {
+          dir("${BASE_DIR}") {
+            withTestClusterEnv {
+              withGoEnv(version: readFile(file: ".go-version").trim()) {
```
Comment (v1v) [inline .ci/smoke-tests.groovy:69]:
```suggestion
              withGoEnv() {
```

File: .ci/smoke-tests.groovy [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  }
+  parameters {
+    string(name: 'SMOKETEST_VERSIONS', defaultValue: 'latest', description: 'Run smoke tests using following APM versions')
+  }
+  stages {
+    stage('Checkout') {
+      options { skipDefaultCheckout() }
+      steps {
+        deleteDir()
+        gitCheckout(basedir: "${BASE_DIR}", shallow: false)
```
Comment (v1v) [inline .ci/smoke-tests.groovy:33]:
```suggestion
        gitCheckout(basedir: "${BASE_DIR}", shallow: false)
        setEnvVar('GO_VERSION', readFile("${BASE_DIR}/.go-version").trim())
```

**Review by v1v — COMMENTED** [review]
A few comments, but would you mind splitting this PR in two, so the `.ci/jobs/**` are in a different PR, so we can easily merge and test whether the pipeline in this particular PR works as expected?

`.ci/jobs` are quite static, so jobs are not generated in the CI as long as they are not merged, while `.ci/*.groovy` are more dynamic as long as the Jenkins Multibranch Job has been created.

File: .ci/smoke-tests.groovy [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  environment {
+    REPO = 'apm-aws-lambda'
+    BASE_DIR = "src/github.com/elastic/${env.REPO}"
+    AWS_ACCOUNT_SECRET = 'secret/observability-team/ci/elastic-observability-aws-account-auth'
+    EC_KEY_SECRET = 'secret/observability-team/ci/elastic-cloud/observability-pro'
+    TERRAFORM_VERSION = '1.2.3'
+    CREATED_DATE = "${new Date().getTime()}"
+    SLACK_CHANNEL = "#apm-serverless"
+  }
+
```
Comment (v1v) [inline .ci/smoke-tests.groovy:15]: To bypass an existing issue while using params for the very first build, this should avoid an empty/unset env variable in those cases

```suggestion
    SMOKETEST_VERSIONS = "${params.SMOKETEST_VERSIONS}"
  }
```

File: .ci/smoke-tests.groovy [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  parameters {
+    string(name: 'SMOKETEST_VERSIONS', defaultValue: 'latest', description: 'Run smoke tests using following APM versions')
+  }
+  stages {
+    stage('Checkout') {
+      options { skipDefaultCheckout() }
+      steps {
+        deleteDir()
+        gitCheckout(basedir: "${BASE_DIR}", shallow: false)
+        setEnvVar('GO_VERSION', readFile("${BASE_DIR}/.go-version").trim())
```
Comment (kruskall) [inline .ci/smoke-tests.groovy:34]: This was actually copied over from smoke test jobs in other repository. Is there any value in using it in this repository ?

I think we're always targeting the latest version of Go and the user should not need to compile the extension.

Comment (v1v) [reply .ci/smoke-tests.groovy:34]: it's used by the `withGoEnv` step implicitly, although the same logic happens internally, `withGoEnv` calls https://github.com/elastic/apm-pipeline-library/blob/4bf7553413212e90a142c649ebb2ebac7b4a5503/vars/goDefaultVersion.groovy#L25-L33

you can remove it if you prefer, though it works as is and it's harmless.

File: .ci/jobs/apm-aws-lambda-smoketests-mbp.yml (hunk 1 above)
Comment (v1v) [reply .ci/jobs/apm-aws-lambda-smoketests-mbp.yml]: > I've moved this file to another PR. Good question on the commit/PRs run.

Let me know when I can review it. The configuration for the `.ci/jobs/*.yml` might require some further analysis depending when those jobs should be ran, as I pointed earlier, happy to discuss further :)

File: .ci/smoke-tests.groovy (hunk 6 above)
Comment (kruskall) [reply .ci/smoke-tests.groovy]: Removed :+1:

File: .ci/jobs/apm-aws-lambda-smoketests-mbp.yml (hunk 1 above)
Comment (kruskall) [reply .ci/jobs/apm-aws-lambda-smoketests-mbp.yml]: Opened a new PR and removed the option :+1:

**Review by v1v — APPROVED** [review]
LGTM, but let's wait for https://github.com/elastic/apm-aws-lambda/pull/329 so we can test this PR without merging it

**Comment by v1v** [comment]
I've just added the support to run always when pushing into the main branch by default. And also support for running on PR basis by using the Jenkins UI.

For instance, this particular PR can be validated manually by running the job -> https://apm-ci.elastic.co/job/library/job/apm-aws-lambda-smoketests-mbp/view/change-requests/job/PR-310/

**Comment by v1v** [comment]
> It seems the job is failing because the aws provider is not configured (missing credentials), could you help with that ? I'm not familiar with the process/guidelines regarding adding credentials to jenkins.

Everything is in Vault, as far as I know it uses the secrets defined in `secret/observability-team/ci/elastic-observability-aws-account-auth`. https://github.com/elastic/apm-aws-lambda/pull/310/files#diff-377731352d232f58c31c9af00e5e8dbbf25eb9a2db0edb045a9853ab9d99fd99R115 is the section to configure AWS accordingly with the command [`aws configure import csv file`](https://github.com/elastic/apm-pipeline-library/blob/main/vars/withAWSEnv.groovy#L59)


What's the error? Maybe there are some missing permissions for the existing AWS account?

**Comment by kruskall** [comment]
> What's the error? Maybe there are some missing permissions for the existing AWS account?

```
[2022-10-12T12:14:35.035Z] Error: error configuring Terraform AWS Provider: failed to get shared config profile, observability-robots@elastic.co
 
[2022-10-12T12:14:35.035Z] 
[2022-10-12T12:14:35.035Z]   with provider["registry.terraform.io/hashicorp/aws"],
[2022-10-12T12:14:35.035Z]   on main.tf line 1, in provider "aws":
[2022-10-12T12:14:35.035Z]    1: provider "aws" {
```

See https://apm-ci.elastic.co/blue/organizations/jenkins/library%2Fapm-aws-lambda-smoketests-mbp/detail/PR-310/5/pipeline

**Comment by v1v** [comment]
I'm not much familiar with that error, I just searched and found:

> If I downgrade to version = "~> 3.0" it fixes the problem. But I would like to upgrade to the latest version.

In [here](https://stackoverflow.com/questions/73055392/error-configuring-terraform-aws-provider-failed-to-get-shared-config-profile)

I also see some references about some specific terraform configuration:
- https://github.com/hashicorp/terraform-provider-aws/issues/15361#issuecomment-699175180
- https://registry.terraform.io/providers/hashicorp/aws/latest/docs#shared-configuration-and-credentials-files

I'm not much familiar with the existing terraform/aws configuration for this project, that particular `withAWSEnv` steps  works in some other projects, do you happen to know what's the `profile` meaning in terms of terraform?

File: .ci/smoke-tests.groovy [hunk 7]
```diff
@@ -0,0 +1,122 @@
+#!/usr/bin/env groovy
+@Library('apm@current') _
```
Comment (v1v) [inline .ci/smoke-tests.groovy:2]: I'm testing out https://github.com/elastic/apm-pipeline-library/pull/1927

```suggestion
@Library('apm@feature/aws-shared-credentials') _
```

**Pull request merged**